File: 2018/02.py

import math, copy, re, hashlib
import itertools as it


def part_1(data):

	twos = 0
	threes = 0
	for i in data:
		has2 = False
		has3 = False
		for l in i:
			if list(i).count(l) == 3:
				has3 = True
				continue
			if list(i).count(l) == 2:
				has2 = True
		threes += 1 if has3 else 0
		twos += 1 if has2 else 0


	print(twos * threes)
	print('END OF PART1')
	return

def get_diff(s1, s2):
	diff = 0
	s = ""
	for i in range(len(s1)):
		if s1[i] != s2[i]:
			diff += 1
		else:
			s += s1[i]
	return (diff, s)

def part_2(data):
	found = False
	res =  None
	for i in range(len(data)):
		for j in range(len(data)):
			if i == j: continue
			diff, string = get_diff(data[i], data[j])
			if diff == 1:
				res = string
				found = True
				break
		if found: break
	print(res)
	print('END OF PART2')
	return 


if __name__ == '__main__':
	with open('02_input') as f:
		data = f.read()
		data = data.split('\n')
		# data = list(map(int, data.split()))


	part_1(copy.deepcopy(data))
	part_2(copy.deepcopy(data))
	

File: 2018/02.js

var fs = require('fs');
const { ord, chr, debug, disp, int, float, randint, gcd, lcm, crt, modPow, mod } = require("../lib");
const { range, drange, trange, iter, diter, titer } = require("../lib");
const { PriorityQueue, Queue, set, Stack } = require("../lib");
const { min, max, random, abs, ceil, floor, log, log10, log2, round, sign, sin, cos, tan, asin, acos, atan, atan2, sqrt, PI } = Math;
const { isSuperset, or, and, xor, sub } = set;

const sort = (arr, fn = (a, b) => a-b) => {
	arr.sort(fn);
}
const in_ = (a, arr) => arr.indexOf(a) !== -1;

const D4 = [[0,1],[1,0],[0,-1],[-1,0]];
const D8 = [[0,1],[1,1],[1,0],[1,-1],[0,-1],[-1,-1],[-1,0],[-1,1]];
const MOD = 1e9+7;

function part1(data) {

	let cnt2 = 0, cnt3 = 0;

	iter(data,i => {
		let arr = new Array(26).fill(0);
		iter(i,j => {
			arr[ord(j) - ord('a')]++;
		});
		let has2 = false, has3 = false;
		iter(arr,j => {

			if (j === 2 && !has2) {
				cnt2++;
				has2 = true;
			}
			if (j === 3 && !has3) {
				cnt3++;
				has3 = true;
			}
		});
	})
	debug(cnt2 * cnt3)
	console.log("END OF PART1");
	return;
}

const diff = (a, b) => {
	let cnt = 0;
	range(a.length)(i => {
		if (a[i] !== b[i]) {
			cnt++;
		}
	})
	return cnt;
}

function part2(data) {

	diter(data,(i, j) => {
		if (i !== j && diff(i,j) === 1) {
			const s1 = new set(i), s2 = new set(j);
			debug([...and(s1,s2)].join(''));
			return true;
		}
	})

	console.log("END OF PART2");
	return;
}

function main() {
	let data = fs.readFileSync("02_input").toString("utf-8");
	data = data.split('\n');
	// data = data.split('\n').map(a => Number(a));

	if (Array.isArray(data)) {
		part1(Array.from(data));
		part2(Array.from(data));
	} else {
		part1(data);
		part2(data);
	}
}

main();
